codegen/masm/src/codegen/scheduler.rs

```diff
@@ -0,0 +1,1404 @@
+    /// that aren't ultimately used due to a runtime condition.
+    pub post: SmallSet<NodeId, 4>,
+    /// The successor blocks and argument count for this instruction
+    pub successors: SmallVec<[Successor; 2]>,
+}
+impl SparseMapValue<hir::Inst> for InstInfo {
+    fn key(&self) -> hir::Inst {
+        self.inst
+    }
+}
+impl InstInfo {
+    #[inline]
+    pub const fn arity(&self) -> usize {
+        self.arity as usize
+    }
+
+    /// Get the constraints for the plain arguments of this instruction
+    pub fn plain_arguments(&self) -> &[Constraint] {
+        &self.args[..self.arity()]
+    }
+
+    /// Get the constraints for the arguments of successor `block`
+    pub fn block_arguments(&self, block: hir::Block) -> &[Constraint] {
+        let range = self.block_argv_range(block);
+        &self.args[range]
+    }
+
+    fn block_argv_range(&self, block: hir::Block) -> core::ops::Range<usize> {
+        let mut start_idx = self.arity();
+        let mut argc = 0;
+        for successor in self.successors.iter() {
+            if successor.block == block {
+                argc = successor.argc as usize;
+                break;
+            }
+            start_idx += successor.argc as usize;
+        }
+        start_idx..(start_idx + argc)
+    }
+}
+
+/// [BlockInfo] describes information about a block relevant to scheduling
+/// and code generation. Namely, it provides access to the computed dependency
+/// graph and tree graph used to schedule the block; but it also provides
+/// convenient access to commonly-queried block information.
+#[derive(Debug)]
+pub struct BlockInfo {
+    /// The source HIR block this info is based on
+    pub source: hir::Block,
+    /// The target MASM block which will be emitted from this info
+    pub target: masm::BlockId,
+    /// The id of the last instruction in the source HIR block,
+    /// this is commonly used to check for liveness after the end
+    /// of a block
+    pub last_inst: hir::Inst,
+    /// The innermost loop to which this block belongs
+    pub innermost_loop: Option<Loop>,
+    /// If set, indicates that this block is the loop header
+    /// for the specified loop.
+    pub loop_header: Option<Loop>,
+    /// The dependency graph of this block
+    pub depgraph: DependencyGraph,
+    /// The topologically-ordered tree graph of this block
+    pub treegraph: OrderedTreeGraph,
+}
+impl BlockInfo {
+    #[inline(always)]
+    pub fn is_loop_header(&self) -> bool {
+        self.loop_header.is_some()
+    }
+}
+impl SparseMapValue<hir::Block> for BlockInfo {
+    fn key(&self) -> hir::Block {
+        self.source
+    }
+}
+
+/// [Schedule] describes an instruction scheduling plan for a single IR function.
+///
+/// The plan describes the order in which blocks will be scheduled,
+/// and the schedule for instructions in each block.
+#[derive(Debug)]
+pub struct Schedule {
+    pub schedule: SmallVec<[hir::Block; 4]>,
```

Comment:
Ah, you're right, didn't catch that since it's `pub`, so no warnings. Originally it was meant to reflect the order in which the blocks were scheduled, but that turned out to not be relevant